internal/graph/cursors.go

```diff
@@ -0,0 +1,287 @@
+		return "", nil
+	}
+
+	if ci.currentCursor.Sections[0] != name {
+		return "", spiceerrors.MustBugf("expected cursor section %s in %v", name, ci.currentCursor.Sections)
+	}
+
+	return ci.currentCursor.Sections[1], nil
+}
+
+// integerSectionValue returns the *integer* found after the `name` at the head of the incoming cursor.
+// If the incoming cursor is empty, returns 0. If the incoming cursor does not start with the name,
+// fails with an error.
+func (ci cursorInformation) integerSectionValue(name string) (int, error) {
+	valueStr, err := ci.sectionValue(name)
+	if err != nil {
+		return 0, err
+	}
+	if valueStr == "" {
+		return 0, err
+	}
+
+	return strconv.Atoi(valueStr)
+}
+
+// mustWithOutgoingSection returns cursorInformation updated with the given name and optional
+// value(s) appended to the PostExecutionCursorSections for the current cursor. If the current
+// cursor already begins with the given name, its value is replaced.
+func (ci cursorInformation) mustWithOutgoingSection(name string, values ...string) cursorInformation {
+	hasSection, err := ci.hasPrefix(name)
+	if err != nil {
+		panic(err)
+	}
+
+	ocs := slices.Clone(ci.outgoingCursorSections)
+	ocs = append(ocs, name)
+	ocs = append(ocs, values...)
+
+	if hasSection {
+		// If the section already exists, remove it and its values in the cursor.
+		return cursorInformation{
+			currentCursor: &v1.Cursor{
+				AtRevision: ci.revision.String(),
+				Sections:   slices.Clone(ci.currentCursor.Sections[len(values)+1:]),
+			},
+			outgoingCursorSections: ocs,
+			limits:                 ci.limits,
+			revision:               ci.revision,
+		}
+	}
+
+	return cursorInformation{
+		currentCursor:          nil,
+		outgoingCursorSections: ocs,
+		limits:                 ci.limits,
+		revision:               ci.revision,
+	}
+}
+
+func (ci cursorInformation) clearIncoming() cursorInformation {
+	return cursorInformation{
+		currentCursor:          nil,
+		outgoingCursorSections: ci.outgoingCursorSections,
+		limits:                 ci.limits,
+		revision:               ci.revision,
+	}
+}
+
+type cursorHandler func(c cursorInformation) error
+
+// withIterableInCursor executes the given handler for each item in the items list, skipping any
+// items marked as completed at the head of the cursor and injecting a cursor representing the current
+// item.
+//
+// For example, if items contains 3 items, and the cursor returned was within the handler for item
+// index #1, then item index #0 will be skipped on subsequent invocation.
+func withIterableInCursor[T any](
+	ci cursorInformation,
+	name string,
+	items []T,
+	handler func(ci cursorInformation, item T) error,
+) error {
+	// Check the index for the section in the cursor. If found, we skip any items before that index.
+	afterIndex, err := ci.integerSectionValue(name)
+	if err != nil {
+		return err
+	}
+
+	isFirstIteration := true
+	for index, item := range items {
+		if index < afterIndex {
```

Comment:
I'm having a hard time understanding why `cursorInformation` has the structure it has. I'm probably missing something since this PR is massive and I'm going commit by commit, but I keep asking myself why the choice of modeling sections and values with a slice. I recall you mentioning maps wouldn't work because we don't have a guarantee of the order, which is very important for the cursor, but why couldn't create a struct field for each section, and give the cursor a more type-safe shape. This kind of index contortions and jumping around is what preoccupies me, if we get anything wrong... we could be skipping elements during the iteration.

Reply:
Because of the arbitrary nesting of cursor sections. As an example, entrypoint #0 might contain a data access underneath it, while entrypoint #1 contains a direct match return, and entrypoint #2 follows a TTU that leads to yet more entrypoints. The chain of sections is *different* for each path of code being followed.